File: android/src/main/java/expo/modules/openwearables/OpenWearablesModule.kt
package expo.modules.openwearables

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.functions.Coroutine
import com.openwearables.health.sdk.*

public class OpenWearablesModule : Module() {
  public override fun definition() = ModuleDefinition {
    Name("OpenWearablesHealthSDK")

    // MARK: - Callbacks (Events)
    Events("onLog", "onAuthError")

    // MARK: - Lifecycle
    OnCreate {
      OpenWearablesHealthSDK.initialize(appContext.reactContext!!)
      
      appContext.activityProvider?.currentActivity?.let { activity ->
        OpenWearablesHealthSDK.getInstance().setActivity(activity)
      }

      OpenWearablesHealthSDK.getInstance().logListener = { message ->
        sendEvent(
          "onLog",
          mapOf(
            "message" to message
          )
        )
      }

      OpenWearablesHealthSDK.getInstance().authErrorListener = { statusCode, message ->
        sendEvent(
          "onAuthError",
          mapOf(
            "statusCode" to statusCode,
            "message" to message
          )
        )
      }
    }

    OnActivityEntersForeground {
      appContext.activityProvider?.currentActivity?.let { activity ->
        OpenWearablesHealthSDK.getInstance().setActivity(activity)
      }

      OpenWearablesHealthSDK.getInstance().onForeground()
    }

    OnActivityEntersBackground {
      OpenWearablesHealthSDK.getInstance().onBackground()
    }

    OnActivityDestroys {
      OpenWearablesHealthSDK.getInstance().unregisterPermissionLauncher()
    }

    // MARK: - Configure
    Function("configure") { host: String, customSyncURL: String? ->
      OpenWearablesHealthSDK.getInstance().configure(host, customSyncURL)
    }

    // MARK: - Auth
    AsyncFunction("signIn") Coroutine { userId: String, accessToken: String?, refreshToken: String?, apiKey: String? ->
      OpenWearablesHealthSDK.getInstance().signIn(
        userId,
        accessToken,
        refreshToken,
        apiKey
      )
    }

    AsyncFunction("signOut") Coroutine { _: Unit? ->
      OpenWearablesHealthSDK.getInstance().signOut()
    }

    Function("updateTokens") { accessToken: String, refreshToken: String ->
      OpenWearablesHealthSDK.getInstance().updateTokens(
        accessToken,
        refreshToken
      )
    }

    Function("restoreSession") {
      OpenWearablesHealthSDK.getInstance().restoreSession()
    }

    Function("isSessionValid") {
      OpenWearablesHealthSDK.getInstance().isSessionValid()
    }

    // MARK: - Authorization
    AsyncFunction("requestAuthorization") Coroutine { types: List<String> ->
      return@Coroutine OpenWearablesHealthSDK.getInstance().requestAuthorization(types)
    }

    // MARK: - Sync
    Function("setSyncInterval") { minutes: Long ->
      OpenWearablesHealthSDK.getInstance().setSyncInterval(minutes)
    }

    AsyncFunction("startBackgroundSync") Coroutine { syncDaysBack: Int? ->
      return@Coroutine OpenWearablesHealthSDK.getInstance().startBackgroundSync(syncDaysBack)
    }

    AsyncFunction("stopBackgroundSync") Coroutine { _: Unit? ->
      OpenWearablesHealthSDK.getInstance().stopBackgroundSync()
    }

    AsyncFunction("syncNow") Coroutine { _: Unit? ->
      OpenWearablesHealthSDK.getInstance().syncNow()
    }

    AsyncFunction("resumeSync") Coroutine { _: Unit? ->
      OpenWearablesHealthSDK.getInstance().resumeSync()
      return@Coroutine true
    }

    Function("isSyncActive") {
      OpenWearablesHealthSDK.getInstance().isSyncActive()
    }

    Function("getSyncStatus") {
      OpenWearablesHealthSDK.getInstance().getSyncStatus()
    }

    Function("resetAnchors") {
      OpenWearablesHealthSDK.getInstance().resetAnchors()
    }

    Function("getStoredCredentials") {
      OpenWearablesHealthSDK.getInstance().getStoredCredentials()
    }

    // MARK: - Providers
    Function("getAvailableProviders") {
      OpenWearablesHealthSDK.getInstance().getAvailableProviders()
    }

    Function("setProvider") { providerId: String ->
      OpenWearablesHealthSDK.getInstance().setProvider(providerId)
    }
    
    // MARK: - Logs
    Function("setLogLevel") { levelId: Int ->
      val level = OWLogLevel.entries.getOrElse(levelId) {
          OWLogLevel.DEBUG
      }
      OpenWearablesHealthSDK.getInstance().logLevel = level
    }

    Function("getLogLevel") {
      OpenWearablesHealthSDK.getInstance().logLevel.ordinal
    }    
  }
}

File: ios/OpenWearablesModule.swift
import ExpoModulesCore
import OpenWearablesHealthSDK

public class OpenWearablesModule: Module {
    public func definition() -> ModuleDefinition {
        Name("OpenWearablesHealthSDK")
        
        // MARK: - Callbacks (Events)        
        Events("onLog", "onAuthError")
        
        // MARK: - Lifecycle        
        OnCreate {
            OpenWearablesHealthSDK.shared.onLog = { message in
                self.sendEvent("onLog", [
                    "message": message
                ])
            }
            
            OpenWearablesHealthSDK.shared.onAuthError = { statusCode, message in
                self.sendEvent("onAuthError", [
                    "statusCode": statusCode,
                    "message": message
                ])
            }
        }
        
        // MARK: - Configure        
        Function("configure") { (host: String, customSyncURL: String?) in
            OpenWearablesHealthSDK.shared.configure(host: host)
        }
        
        // MARK: - Auth        
        AsyncFunction("signIn") { (
            userId: String,
            accessToken: String?,
            refreshToken: String?,
            apiKey: String?,
            promise: Promise
        ) in
            OpenWearablesHealthSDK.shared.signIn(
                userId: userId,
                accessToken: accessToken,
                refreshToken: refreshToken,
                apiKey: apiKey
            )
            promise.resolve()
        }
        
        AsyncFunction("signOut") { (promise: Promise) in
            OpenWearablesHealthSDK.shared.signOut()
            promise.resolve()
        }
        
        Function("updateTokens") { (accessToken: String, refreshToken: String) in
            OpenWearablesHealthSDK.shared.updateTokens(
                accessToken: accessToken,
                refreshToken: refreshToken
            )
        }
        
        Function("restoreSession") {
            return OpenWearablesHealthSDK.shared.restoreSession()
        }
        
        Function("isSessionValid") {
            return OpenWearablesHealthSDK.shared.isSessionValid
        }
        
        // MARK: - HealthKit Authorization        
        AsyncFunction("requestAuthorization") { (types: [String], promise: Promise) in
            let healthTypes = types.compactMap { HealthDataType(rawValue: $0) }
            OpenWearablesHealthSDK.shared.requestAuthorization(types: healthTypes) { success in
                promise.resolve(success)
            }
        }
        
        // MARK: - Sync    
        Function("setSyncInterval") { (minutes: Double) in } // (not implemented in iOS SDK)
            
        AsyncFunction("startBackgroundSync") { (syncDaysBack: Int?, promise: Promise) in
            OpenWearablesHealthSDK.shared.startBackgroundSync(syncDaysBack: syncDaysBack) { started in
                promise.resolve(started)
            }
        }
        
        AsyncFunction("stopBackgroundSync") { (promise: Promise) in
            OpenWearablesHealthSDK.shared.stopBackgroundSync()
            promise.resolve()
        }
        
        AsyncFunction("syncNow") { (promise: Promise) in
            OpenWearablesHealthSDK.shared.syncNow {
                promise.resolve()
            }
        }
        
        AsyncFunction("resumeSync") { (promise: Promise) in
            OpenWearablesHealthSDK.shared.resumeSync { resumed in
                promise.resolve(resumed)
            }
        }
        
        Function("isSyncActive") {
            return OpenWearablesHealthSDK.shared.isSyncActive
        }
        
        Function("getSyncStatus") {
            return OpenWearablesHealthSDK.shared.getSyncStatus()
        }
        
        Function("resetAnchors") {
            OpenWearablesHealthSDK.shared.resetAnchors()
        }
        
        Function("getStoredCredentials") {
            return OpenWearablesHealthSDK.shared.getStoredCredentials()
        }

        // MARK: - Providers (not implemented in iOS SDK)        
        Function("getAvailableProviders") { return [] }
        
        Function("setProvider") { }

        // MARK: - Logs
        Function("setLogLevel") { (levelId: Int) in
            let level = OWLogLevel(rawValue: levelId) ?? .debug
            OpenWearablesHealthSDK.shared.setLogLevel(level)
        }

        Function("getLogLevel") {
            return OpenWearablesHealthSDK.shared.logLevel.rawValue
        }
    }
}